docs/decisions/0011-enterprise-subsidy-enrollments-and-entitlements.rst

```diff
@@ -0,0 +1,143 @@
+## Enterprise subsidy enrollments and entitlements
+
+# Status
+
+Draft
+
+# Context
+
+Enterprise course enrollment and course entitlement records can come into existence from different types of subsidies:
+
+- Subscription licenses - we’d want to track the license uuid at time of enrollment creation.
+
+- Learner credit subsidy transactions - we’d want to track the transaction uuid at time of enrollment or entitlement creation.
+
+- Coupon codes - not yet tracked in the context of this ADR.
+
+On top of architecting the tables to support sub licenses and credit transactions we want to build a system that can support other, not yet known types of subsidies as there is the chance that new ones will be added in the future.
+
+We also will very likely need to create a new EnterpriseCourseEntitlement model, which would mirror the purpose of the EnterpriseCourseEnrollment model but instead track a subsidy that is not yet connected to a course enrollment and will support upgrading to said enrollment.
+
+**Why implement table inheritance and what is `EnterpriseFulfillmentSource`?**
+
+- Having concrete table inheritance (`EnterpriseFulfillmentSource` being the parent model of all enterprise related subsidy based enrollments) means that we have a single entry point table that every kind of transaction/enrollment/entitlement/etc will extend from.
```

Comment:
We considered doing concrete Django model inheritance, but instead used abstract inheritance because the complexity required to do data and schema migrations was too great to justify the costs and risks.  Abstract inheritance still gives us the benefits of:
1. flexibility to have subsidy-specific methods and fields on the child models.
2. common behavior of methods via the base class.
3. common and unique `uuid` field for easy reference across systems (e.g. the uuid is what's stored on ledger transactions to commit them).

Mitigation for reporting/analytics: can easily write a view to aggregate these two tables into a single relation with e.g. a type column to easily distinguish between subsidy types.